src/ext/respToTcpLoader.c

```diff
@@ -9,72 +9,125 @@
 #endif
 
-#define NUM_PENDING_CMDS_HIGH_THRESHOLD 200
-#define NUM_PENDING_CMDS_LOW_THRESHOLD  100
+#define NUM_PENDING_CMDS_HIGH_THD 200
+#define NUM_PENDING_CMDS_LOW_THD  100
+
+#define NUM_RECORDED_CMDS         400
+#define RECORDED_MAX_SIZE         40
+#define REPLY_BUFF_SIZE           4096
+
+#define INITIAL_DELAY_USEC        20
+#define MAX_RETRIES               6
 
 struct RdbxRespToTcpLoader {
 
     struct {
         int num;
         int highThreshold;
         int lowThreshold;
+
+        char cmdPrefix[NUM_RECORDED_CMDS][RECORDED_MAX_SIZE];
+        int cmdAt;
     } pendingCmds;
 
+    struct {
+        size_t nResponses;
+        size_t nErrors;
+    } protocol;
+
     size_t sentCmds;
 
     RdbParser *p;
     int fd;
 };
 
 /* return 0 for success. 1 Otherwise. */
-int readReplies(RdbxRespToTcpLoader *ctx, int numToRead) {
-    char buffer[2048];
-    int numRead = 0;
-
-    while (numRead < numToRead) {
-        /* Read a single-line reply into the buffer */
-        ssize_t bytesRead = read(ctx->fd, buffer, sizeof(buffer) - 1);
-
-        /* TODO: Check responses for errors. */
-        /* TODO: Record pending commands in case of failure */
-        if (bytesRead > 0) {
-            /* Iterate through the read bytes to find the end of the lines */
-            for (int i = 0; i < bytesRead; i++) {
-                if (buffer[i] == '\n') {
-                    numRead++;
-                }
+static int readReplies(RdbxRespToTcpLoader *ctx, int numToRead) {
+    char buff[REPLY_BUFF_SIZE];
+    RespReaderCtx respCtx;
+    readRespInit(&respCtx);
+
+    int retries = 0;
+    int delayUs = INITIAL_DELAY_USEC;
+
+    while ((int)respCtx.countReplies < numToRead) {
+        int rd = recv(ctx->fd, buff, sizeof(buff), MSG_DONTWAIT);
+
+        if (rd > 0) {
+            /* Data was received, process it */
+            if (RESP_REPLY_ERR == readRespReplies(&respCtx, buff, rd)) {
+                RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP_READ, respCtx.errorMsg);
+                return 1;
             }
-        } else {
-            //
-            RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_FAILED_READ,
-                            "Failed to read responses from socket.");
-            ctx->pendingCmds.num -= numRead;
+
+        } else if (rd == 0) {
+            RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_CONN_CLOSE, "Connection closed by the remote side");
             return 1;
+        } else {
+            /* Error occurred or no data available */
+            if (errno == EWOULDBLOCK || errno == EAGAIN) {
+                /* No data available, retry after delay */
+                if (retries >= MAX_RETRIES) {
+                    RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_MAX_RETRIES, "Maximum number of retries reached. Exit.");
+                    return 1;
+                }
+                usleep(delayUs);
+                delayUs *= 2;  /* Double the delay for each retry */
```

Comment:
maybe i'm out of sync with the recent conversation about this (i remember seeing something), or maybe i didn't read all the code yet, but we do plan to get rid of this sleep right?
i.e. we should allow some pipeline of requests until we read responses, but when we eventually decide to block on a response, it should be a blocking IO (not a sleep).

please point me to what i'm missing.

Reply:
I wasn't paying attention as i was thinking that i will get back to this code when doing profiling.
(I also had wrong understanding that blocking IO read() might add wait when it receieves less data than requested)